source/host_resolver.c

```diff
@@ -489,9 +496,13 @@ static void on_host_value_removed(void *value) {
         aws_thread_clean_up(&host_entry->resolver_thread);
     }
 
+    if (!aws_linked_list_empty(&host_entry->pending_resolution_callbacks)) {
+        aws_raise_error(AWS_IO_DNS_HOST_REMOVED_FROM_CACHE);
+    }
+
     while (!aws_linked_list_empty(&host_entry->pending_resolution_callbacks)) {
         struct aws_linked_list_node *resolution_callback_node =
-            aws_linked_list_front(&host_entry->pending_resolution_callbacks);
+            aws_linked_list_pop_front(&host_entry->pending_resolution_callbacks);
```

Comment:
How did this ever work?!

Reply:
Because the path was never being hit, due to the fact that it was impossible for host resolver to be cleaned up, because channel bootstrap wasn't cleaned up until all pending resolutions were finished.